### ISSUE TITLE
Question: GPU acceleration

### ISSUE THREAD

**Issue by NowayIndustries** [inspirehep/magpie#96]
Hello again,

Just a quick question, has someone already tried using GPU acceleration with Magpie?

According to my searches online I would assume that because of the several layers of abstraction tensorflow/theano & keras) that this would be trivial to implement. And many people describe quite significant increases in training speed (I've seen reports from 4x to 60x), altough this increase is heavily dependent upon the neural network being used.

For the project I am working on it would be nice to be able train (several) models every night, to update the labels that can be predicted. Currently my corpus (that is probably a bit on the big side for the use case) takes between two and three hours at 1100% CPU usage.

So to make sort of a tl;dr:

- Is it possible to train Magpie's neural network on GPU's?
- Have you tried it? What kind of speed improvements have you seen?
- Do you think trying it on an AWS p2 instance (or something similair) will provide a 'real-world' case of performance increase (so I can justify the spending, should I have to)

**Comment by jstypka** [comment]
Yes, I used it with a GPU. So when you install Tensorflow/Theano you need to set it up to work with a GPU: https://www.tensorflow.org/install/install_linux and then Magpie uses it out-of-the-box. The speedup depends on the usecase, but in my case it was pretty big.

Depending on the GPU there are some libraries that need to be installed, but all of them are free as far as I recall. But then again, you do it while setting up TensorFlow - check out the link that I posted.

**Comment by jstypka** [comment]
I don't know a lot about AWS, but I wouldn't be surprised if they provide nodes with GPUs, all the libraries and TensorFlow installed, so you would just put Magpie on top and let it fly

**Comment by tarvos21** [comment]
@NowayIndustries hi, I have tried to use Magpie on Google Cloud Platform to classify text with thousands of labels, I choose an instance of 8CPU, 40GB RAM and a Nvidia Tesla K80 GPU. I installed Tensorflow-gpu, and Magpie will use it automatically, as far as I can see, it is very convenient for training.
I also have a 2015 MacBook Pro for coding and training NN with magpie, and I conclude that a Nvidia Tesla K80 with 12GB VRAM is 6 times faster than the MacBook Pro which have no available Nvidia GPU to use.
I have also checked the p2.8xlarge/16xlarge instances on AWS, they also uses K80/K40 GPU, I am sure it's almost the same situation on Google Cloud Platform.

**Comment by NowayIndustries** [comment]
Hello everyone,
Thank you for your information, and answering my questions.

Seems like it really influences training times, which will be great for the production environment.
If/when I get some performance metrics I will add them to this issue aswell, so there can be some sort documentation on GPU accelerated magpie performance.

For now though I'll close the issue because I've received my answers, altough more performance data will always be better (right?).
